---
title: "RIBOLOG: General description and workflow"
author: "Hosseinali Asgharian"
date: "Sep. 2019"
output: rmarkdown::html_document
vignette: >
  %\VignetteIndexEntry{RIBOLOG: General description and workflow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
library(knitr)
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
opts_chunk$set(root.dir = "C:/Science/Projects/Ribosome profiling/Ribolog")
```

Ribolog includes tools to perform a variety of analyses based on ribosome profiling data. Module 1: CELP (Consistent Excess of Loess Preds) identifies positions of translational pause (stalling) and corrects RPF counts to eliminate the impact of stalling bias. The output of CELP can be also used to model the factors that influence translational dynamics. Module 2: PREP normalizes and combines RNA and RPF datasets and shapes them into a format ready for quality control (QC) and translational efficiency ratio (TER) anlaysis. Module 3: QC includes three powerful tools to quantify and visualize reproducibility among replicates and inform hypothesis generation with respect to biological effects on translation. One is princiapl component analysis (PCA) which can be done on RPF counts, RNA counts or translational efficiencies (RPF/RNA). Two novel tests involve proportion of null features (not-differentially translated transcripts) and correlation of equivalent TER tests in rep-by-rep comparisons. Module 4: TER tests the size and significance of differential translation rates among biological samples. Although better results are always obtained with sufficient replicates, Ribolog is still able to peform the TER test with only one replicate per sample. The TER test is not restricted to pairwise comparisons; complex experimental designs including any number of samples described by several indenpedent variables can be incorporated into a single model (both univariate and multivariate models are supported). The chart below demonstrates the typical Ribolog workflow. Details of each step and available options are described in the vignettes of individual modules and in function documentations.

```{r pressure, echo=FALSE, fig.cap="The Ribolog workflow. First, bam files are read in. RNA and RPF reads per transcript are counted. RPF counts can be optionally corrected to remove stalling bias using the CELP method (module 1). RNA and RPF counts are then normalized for library size variation separately before or after merging (module 2). QC tests are run on the combined RNA-RPF data set to decide on the cutoffs to filter out low-count transcripts or bad samples (module 3). Some trial and error is usually needed until a satisfactorily reliable data set is obtained. This cleaned-up data set is input to module 4 to compare tranlational efficiency ratio among samples and test the effect of biological factors (independent variables) on translation. Numbers 1-4 in dark blue rectangles refer to the module containing each operation. Grey boxes show steps that are performed outside Ribolog using other appropriate tools.", out.width = '100%', fig.align='left'}
knitr::include_graphics("C:/Science/Projects/Ribosome profiling/Ribolog/data-raw/Ribolog_workflow.v4.png")
```

Ribolog is still a work in progress. Four other modules are being prepared and will be released in near future. Feel free to contact us about issues with current functions and modules, or with special requests or ideas about additional analyses you would like to be able to do with ribosome profiling data. 
