---
title: "Ribolog benchmark"
author: "Hossein Asgharian; Sohit Miglani; Hani Goodarzi"
date: "Sep. 2022"
output: html_document
---

```{r setup, include = FALSE}
library(knitr)
opts_chunk$set(root.dir = "C:/Science/Projects/Ribosome Profiling/Benchmark/")
opts_knit$set(root.dir = "C:/Science/Projects/Ribosome Profiling/Benchmark/")

```

```{r, include = FALSE}
library(data.table)
library(tidyr)
library(ggplot2)
library(tidyverse)
library(corrr)
library(yardstick)
library(cowplot)
#library(Ribolog)
#library(kableExtra)
```


In this document, we compare the capabilities and performance metrics of Ribolog with four alternative methods: Xtail, RiboDiff, Riborex and Anota2seq.  
  
## 1. Catalogue of functionalities

Table 1 list the main tests and tools offered by Ribolog and whether they are included by four alternative methods for TER testing and one alternative method for stalling bias characterization. 

```{r, echo = FALSE, results = 'asis'}
table1_functions <- fread("C:/Science/Projects/Ribosome Profiling/Benchmark/Table_functionalities.csv", header = TRUE)
kable(table1_functions, caption = "Table 1. Main capabilities of Ribolog compared with five other ribo-seq packages.") #%>%
 # row_spec(10, extra_css = "border-bottom: 1px solid")
```

This table is not exhaustive. For a more comprehensive review of computational tools for ribosome profiling data visualization, pre-processing and analysis, please see:

Kiniry, S.J., Michel, A.M. and Baranov, P.V., 2020. Computational methods for ribosome profiling data analysis. Wiley Interdisciplinary Reviews: RNA, 11(3), p.e1577.  

  
## 2. Performance metrics and run time

We used the datasets simulated by the authors of Xtail, provided in the supplementary information of the Xiao et al. 2016 paper. The experimental design included two conditions (control and treatment). We ran Ribolog, Xtail, RiboDiff, Riborex and Anota2seq on the 2 vs. 2 and the 3 vs. 3 replicated datasets. Unless explicitly mentioned, all reported results pertain to the 3 vs. 3 comparison.  


```{r, include = FALSE}
setwd("./Hani_5methods/")
ribo_df <- fread("ribolog_3_redo.csv") %>%
           mutate(logTER = lnTER / log(2)) %>%
           dplyr::select(transcript, logTER, pval, padj)
xtail_df <- fread("xtail_3_samples.csv") %>%
            mutate(transcript=Ensembl_ID, logTER=log2FC_TE_final, pval=pvalue_final, padj=pvalue.adjust) %>%
            dplyr::select(transcript, logTER, pval, padj)
rex_df <- fread("riborex_3.csv")%>%
          mutate(logTER = log2FoldChange, pval=pvalue) %>%
          dplyr::select(transcript, logTER, pval, padj)
rdiff_df <- fread("ribdiff_3.tsv") %>%
            mutate(transcript=geneIDs, logTER = `log2FC_TE(DrugTreated vs Control)`) %>%
            dplyr::select(transcript, logTER, pval, padj)
an2s_df <- fread("anota2seq_3.csv") %>%
           mutate(logTER.An2seq=apvEff, pval.An2seq=apvRvmP, padj.An2seq=apvRvmPAdj) %>%
           #mutate(logTER=apvEff, pval=apvRvmP, padj=apvRvmPAdj) %>%
           #dplyr::select(transcript, logTER, pval, padj)
           dplyr::select(transcript, logTER.An2seq, pval.An2seq, padj.An2seq)
  

head(ribo_df)
head(xtail_df)
head(rex_df)
head(rdiff_df)
head(an2s_df)

```

```{r, include = FALSE}
merged_df <- 
ribo_df %>% inner_join(xtail_df, by="transcript", suffix = c(".Ribolog", ".Xtail")) %>%
            inner_join(rex_df  , by="transcript") %>%
            inner_join(rdiff_df, by="transcript", suffix = c(".Riborex", ".Ribodiff")) %>%
            inner_join(an2s_df , by="transcript")
head(merged_df)
```

### 2.1. p-values and TERs

Figure 1 shows that all tested methods generate the expected shape for p-value distribution with a uniform base and a peak towards zero - enriched in differentially translated genes. Tables 2 and 3 demonstrate generally high concordance among all methods except for anota2seq.

```{r, echo = FALSE, message=FALSE, warning=FALSE}
pval_df <- merged_df %>% select(starts_with("pval"))

melt(pval_df) %>% mutate(variable = factor(gsub("pval.", "", variable),c("Ribolog", "Xtail", "Ribodiff", "Riborex", "An2seq"))) %>%
ggplot(aes(x=value, fill=variable)) + ggtitle("Figure 1. Distribution of unadjusted p-values for 3 replicates") +
xlab("p-values in 3 vs 3 comparison") +
  geom_histogram(position="dodge") + theme_bw(11) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), legend.position=c(0.85,0.7)) +
  labs(fill = "Method") +
  scale_fill_manual(values = c("Ribolog" = "#ef476f",
                               "Xtail" = "#ffd166",
                               "Ribodiff" = "#06d6a0",
                               "Riborex" = "#118ab2",
                               "An2seq" = "#073b4c"))
```


```{r, echo = FALSE, message=FALSE, warning=FALSE}
logTER_df <- merged_df %>% select(starts_with("logTER"))
logTER_df_cormat <- logTER_df %>% correlate() %>% as.data.frame()  
row.names(logTER_df_cormat) <- gsub("logTER.", "", logTER_df_cormat$term)
colnames(logTER_df_cormat) <- gsub("logTER.", "", colnames(logTER_df_cormat))
logTER_df_cormat[, -1] %>% kable(caption = "Table 2. Correlation matrix of logTER values by the 5 methods", header = TRUE)
```

```{r, echo = FALSE, message=FALSE, warning=FALSE}
padj_df <- merged_df %>% select(starts_with("padj"))
padj_df_cormat <- padj_df %>% correlate() %>% as.data.frame()  
row.names(padj_df_cormat) <- gsub("padj.", "", padj_df_cormat$term)
colnames(padj_df_cormat) <- gsub("padj.", "", colnames(padj_df_cormat))
padj_df_cormat[, -1] %>% kable(caption = "Table 3. Correlation matrix of adjusted p-values by the 5 methods", header = TRUE)
```

    
### 2.2. Precision-recall and ROC curves

```{r, include = FALSE}
true_df <- fread("./Hani_5methods/truth.csv", header = TRUE) %>% 
    mutate(transcript = Ensembl_ID, target=`Positive(T) or Negative(F)`) %>%
    select(transcript, target)
head(true_df)
```


```{r, include = FALSE}
merge_z <- merged_df %>% inner_join(true_df, by="transcript") %>%
    select(transcript, target, starts_with("pval")) %>%
    pivot_longer(pval.Ribolog:pval.An2seq, names_to = "analysis", values_to="pval")
head(merge_z)
```

```{r, echo = FALSE}
pr_auc_df <- merge_z %>% group_by(analysis) %>% pr_auc(factor(target), pval) %>%
  mutate(analysis = gsub("pval.", "", analysis)) %>%
  rename(Method = analysis, PR_AUC = .estimate) %>%
  dplyr::select(Method, PR_AUC)

roc_auc_df <- merge_z %>% group_by(analysis) %>% roc_auc(factor(target), pval) %>%
  mutate(analysis = gsub("pval.", "", analysis)) %>%
  rename(Method = analysis, ROC_AUC = .estimate) %>%
  dplyr::select(Method, ROC_AUC) 

auc_df <- merge(roc_auc_df,pr_auc_df, by = "Method")

auc_df %>% kable(caption = "Table 4. Receiver operating characteristic curve (ROC) and precision-recall curve AUC for the 5 methods")

```


```{r, echo = FALSE, message = FALSE, warning = FALSE, results='hide'}
p_roc <- merge_z %>% mutate(analysis = factor(gsub("pval.", "", analysis),c("Ribolog", "Xtail", "Ribodiff", "Riborex", "An2seq"))) %>%
   group_by(analysis) %>% roc_curve(factor(target), pval) %>% autoplot() +
  theme_bw(11) + theme(panel.grid.minor = element_blank()) +  coord_equal() +
  scale_color_manual(values = c("Ribolog" = "#ef476f",
                               "Xtail" = "#ffd166",
                               "Ribodiff" = "#06d6a0",
                               "Riborex" = "#118ab2",
                               "An2seq" = "#073b4c")) +
  ggtitle("Figure 2. ROC curve for the 5 methods")

p_roc

p_pr <- merge_z %>% mutate(analysis = factor(gsub("pval.", "", analysis),c("Ribolog", "Xtail", "Ribodiff", "Riborex", "An2seq"))) %>%
   group_by(analysis) %>% pr_curve(factor(target), pval) %>% autoplot() + 
  theme_bw(11) + theme(panel.grid.minor = element_blank()) + coord_equal() +
  scale_color_manual(values = c("Ribolog" = "#ef476f",
                               "Xtail" = "#ffd166",
                               "Ribodiff" = "#06d6a0",
                               "Riborex" = "#118ab2",
                               "An2seq" = "#073b4c")) +
  ggtitle("Figure 3. Precision-recall curve for the 5 methods")

p_pr

#plot_grid(p_roc, p_pr, labels = c("ROC curve", "Precision-recall curve"))
```
   
### 2.3. Robustness to sequencing coverage variation

We randomly subsampled reads to bring the total read count of each sample to 1X, 1/2X, 1/5X and 1/10X of its original total read count. We ran them through the usual analysis pipeline for each method and recorded the performance metrics. The 3 vs 3 tests were run for all methods, and the 2 vs 2 for Ribolog and Xtail. Figures 4-7 show that by sacrificing ~5% specificity, Ribolog retains much of its sensitivity even at extremely low coverages. For example, at 0.1X coverage, sensitivity of 3 vs 3 tests drops from 0.90 to 0.63 for Ribolog, and from 0.83 to 0.29 for Xtail.

```{r,include = FALSE}
sub_df <- fread("./Hani_5methods/sens_spec_table.tsv")
sub_df <- sub_df %>% mutate(sub.sample = factor(sub.sample, levels=c("1X", "2X", "5X", "10X")))
sub_df
```
```{r, echo = FALSE}
p_sub3v3_sens <- sub_df %>% filter(metric=="sensitivity", type=="3v3") %>%
mutate(analysis = factor(analysis,levels=c("Ribolog", "Xtail", "Ribodiff", "Riborex", "Anota2seq"))) %>%
ggplot(aes(x=sub.sample, y=value, fill=analysis)) + geom_bar(stat = "identity", position="dodge") +
xlab("Fold sub-sampled") + ylab("Sensitivity") +
scale_fill_manual(values=c("#ef476f", "#ffd166", "#06d6a0", "#118ab2", "#073b4c"))+
theme_bw(11) + theme(panel.background = element_rect(colour = "black"), panel.grid.minor = element_blank()) + #theme(legend.position = c(0.9, 0.8)) +
  labs(fill = "Method") +
  ggtitle("Figure 4. Retention of sensitivity against decreasing coverage, 3 vs 3 tests.")
p_sub3v3_sens

p_sub2v2_sens <- sub_df %>% filter(metric=="sensitivity", type=="2v2") %>%
mutate(analysis = factor(analysis,levels=c("Ribolog", "Xtail", "Ribodiff", "Riborex", "Anota2seq"))) %>%
ggplot(aes(x=sub.sample, y=value, fill=analysis)) + geom_bar(stat = "identity", position="dodge") +
xlab("Fold sub-sampled") + ylab("Sensitivity") +
scale_fill_manual(values=c("#ef476f", "#ffd166", "#06d6a0", "#118ab2", "#073b4c"))+
theme_bw(11) + theme(panel.background = element_rect(colour = "black"), panel.grid.minor = element_blank()) + #theme(legend.position = c(0.9, 0.8)) +
  labs(fill = "Method") +
  ggtitle("Figure 5. Retention of sensitivity against decreasing coverage, 2 vs 2 tests.")
p_sub2v2_sens

```
```{r, include=FALSE}
#plot_grid(p_sub3v3_sens, p_sub2v2_sens, labels = c("3 vs 3 tests", "2 vs 2 tests"))
```
```{r,echo = FALSE}
p_sub3v3_spec <- sub_df %>% filter(metric=="specificity", type=="3v3") %>%
mutate(analysis = factor(analysis,levels=c("Ribolog", "Xtail", "Ribodiff", "Riborex", "Anota2seq"))) %>%
ggplot(aes(x=sub.sample, y=value, fill=analysis)) + geom_bar(stat = "identity", position="dodge") +
xlab("Fold sub-sampled") + ylab("specificity") +
scale_fill_manual(values=c("#ef476f", "#ffd166", "#06d6a0", "#118ab2", "#073b4c"))+
theme_bw(11) + theme(panel.background = element_rect(colour = "black"), panel.grid.minor = element_blank()) + #theme(legend.position = c(0.9, 0.8)) + 
  labs(fill = "Method") +
  ggtitle("Figure 6. Retention of specificity against decreasing coverage, 3 vs 3 tests.")
p_sub3v3_spec

p_sub2v2_spec <- sub_df %>% filter(metric=="specificity", type=="2v2") %>%
mutate(analysis = factor(analysis,levels=c("Ribolog", "Xtail", "Ribodiff", "Riborex", "Anota2seq"))) %>%
ggplot(aes(x=sub.sample, y=value, fill=analysis)) + geom_bar(stat = "identity", position="dodge") +
xlab("Fold sub-sampled") + ylab("specificity") +
scale_fill_manual(values=c("#ef476f", "#ffd166", "#06d6a0", "#118ab2", "#073b4c"))+
theme_bw(11) + theme(panel.background = element_rect(colour = "black"), panel.grid.minor = element_blank()) + #theme(legend.position = c(0.9, 0.8)) + 
  labs(fill = "Method") +
  ggtitle("Figure 7. Retention of specificity against decreasing coverage, 2 vs 2 tests.")
p_sub2v2_spec

```
   
### 2.4. Run times

Even with the empirical null testing - which improves specificity tremendously but adds to processing time - Ribolog runs as fast as or faster than other methods, except for Riborex.

```{r, include = FALSE}
runtime_df <- fread("./Hani_5methods/cpu_time.tsv")
runtime_df <- runtime_df %>% mutate(analysis = factor(analysis, levels=c("Riborex", "Ribolog", "Ribolog+ENZ", "Anota2seq", "Ribodiff", "Xtail")))
runtime_df

```
```{r, echo = FALSE, fig.width=7, fig.height=6}
runtime_df %>%  filter(type=="3v3") %>% ggplot(aes(x=analysis, y=`cpu.time (sec)`, fill=analysis)) + geom_bar(stat = "identity", position="dodge") +
xlab("") + ylab("CPU time (sec)") +
scale_y_continuous(trans='log10')+
theme_bw(11) + theme(panel.background = element_rect(colour = "black"), panel.grid.minor = element_blank(), axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=0.35)) +
  scale_fill_manual(values=c("#118ab2", "#f37694", "#ef476f", "#073b4c", "#06d6a0", "#ffd166")) +
  ggtitle("Figure 8. Run time of 3 vs 3 tests for Ribolog with and without\nempirical null testing, compared with four alternative methods.\nNote: The y axis is on log scale.") +
  labs(fill = "Method")
```

  
### 2.5. Effect of stalling bias correction with CELP

The CELP (Consistent Excess of Loess Predictions) module of Ribolog detects and corrects stalling bias and other types of local RPF read density nonuniformity in ribo-seq data. To evaluate the impact of this correction, we artificially multiplied the read counts at one codon ("GTC") by 10 in ~300 genes of three MDA cell line samples. Then, we ran Ribolog and Xtail to compare the two groups (original MDA vs GTC-stalled MDA). Uncorrected, these simulated stalling reads were expected to contribute to the total read count in one group of samples and yield inflated TER estimates. Figures 9 and 10 shows that the application of CELP significantly reduces this bias compared with naive Ribolog (without CELP) and Xtail. Note that CELP was applied to the whole dataset here, without any prior knowledge of which genes, which samples or which codons were affected by the simulated bias. 

```{r, include=FALSE}
selected.genes.nz <- fread("./Hani_5methods/modified_gene_list.txt", header = FALSE) %>% rename(transcript = V1)
xtail_mda_gtc <- read.table("./Hani_5methods/xtail_df.tsv") 
xtail_mda_gtc <- xtail_mda_gtc %>%
  mutate(transcript = row.names(xtail_mda_gtc))

ribolog_obs_df <- read.table("./Hani_5methods/ribolog_obs_df.tsv")
ribolog_obs_df <- ribolog_obs_df %>%
  mutate(transcript = row.names(ribolog_obs_df))

ribolog_corr_df <- read.table("./Hani_5methods/ribolog_corr_df.tsv")
ribolog_corr_df <- ribolog_corr_df %>%
  mutate(transcript = row.names(ribolog_corr_df))



dim(selected.genes.nz)
head(selected.genes.nz)

dim(xtail_mda_gtc)
head(xtail_mda_gtc)

#wilcox.test(xtail_mda_gtc[rownames(xtail_mda_gtc) %in% selected.genes.nz,'log2FC_TE_final'])
#wilcox.test(!xtail_mda_gtc[rownames(xtail_mda_gtc) %in% selected.genes.nz,'log2FC_TE_final'])
dim(ribolog_obs_df)
head(ribolog_obs_df)
dim(ribolog_corr_df)
head(ribolog_corr_df)
```

```{r, echo = FALSE}
xt_gtc <- xtail_mda_gtc %>%
  filter(transcript %in% selected.genes.nz$transcript) %>%
  dplyr::select('log2FC_TE_final')

rb.o_gtc <- ribolog_obs_df %>%
  filter(transcript %in% selected.genes.nz$transcript) %>%
  dplyr::select('lnTER') %>%
  mutate(log2TER = lnTER / log(2))

rb.c_gtc <- ribolog_corr_df %>%
  filter(transcript %in% selected.genes.nz$transcript) %>%
  dplyr::select('lnTER') %>%
  mutate(log2TER = lnTER / log(2))

#dim(xt_gtc)
#head(xt_gtc)
#dim(rb.o_gtc)
#head(rb.o_gtc)
#dim(rb.c_gtc)
#head(rb.c_gtc)

merged_gtc <- data.frame(logTER = c(xt_gtc$log2FC_TE_final, rb.c_gtc$log2TER, rb.o_gtc$log2TER),
                     name   = c(rep("Xtail", dim(xt_gtc)[1]),rep("Ribolog (CELP)", dim(rb.c_gtc)[1]),rep("Ribolog naive", dim(rb.o_gtc)[1])))
#head(merged_gtc)

pv.celp = sprintf("P = %.2f", t.test(xt_gtc$log2FC_TE_final, rb.c_gtc$log2TER)$p.value)
pv.rb = sprintf("P = %.2f", t.test(xt_gtc$log2FC_TE_final, rb.o_gtc$log2TER)$p.value)

#pv.celp_paired = sprintf("P = %.2f", t.test(xt_gtc$log2FC_TE_final, rb.c_gtc$log2TER, paired = TRUE)$p.value)
#pv.rb_paired = sprintf("P = %.2f", t.test(xt_gtc$log2FC_TE_final, rb.o_gtc$log2TER, paired = TRUE)$p.value)

#wilcox.test(xt_gtc$log2FC_TE_final, rb.c_gtc$log2TER, paired = TRUE, alternative = "two.sided")
#wilcox.test(xt_gtc$log2FC_TE_final, rb.o_gtc$log2TER, paired = TRUE, alternative = "two.sided")
#wilcox.test(rb.c_gtc$log2TER, rb.o_gtc$log2TER, paired = TRUE, alternative = "two.sided")


#rb.c <- fit_FDR_corr[rownames(fit_FDR_corr) %in% selected.genes.nz,'lnTER'] / log(2)
#rb.o <- fit_FDR_obs[rownames(fit_FDR_obs) %in% selected.genes.nz,'lnTER'] / log(2)
```


```{r, echo=FALSE}
ggplot(merged_gtc) +
  geom_density(aes(logTER, color = name)) +
  labs(color = "Method") + xlab("log2 TER")+
  theme_bw(11) + theme(panel.background = element_rect(colour = "black"), panel.grid.minor = element_blank()) +
  #scale_fill_manual(values=c("#ef476f", "#073b4c", "#ffd166")) +
  scale_color_manual(values=c("#ef476f", "#073b4c", "#ffd166")) +
  ggtitle("Figure 9. Density plots of log2 TER \nbetween original MDA vs GTC-stalled MDA samples.")
  

merged_gtc %>% ggplot(aes(x=name, y=logTER, fill=name)) + geom_boxplot() + ylim(-0.01,0.25) + theme_bw(11) +
#scale_fill_manual(values=c("#FFA6A6", "#A6BDFF", "#9CD39F"))+
ylab("log2 TER") + xlab("Method") +
annotate("text", x=1,y=0.24, label=pv.celp)+
annotate("text", x=2,y=0.24, label=pv.rb)+
theme(panel.background = element_rect(colour = "black"),
      panel.grid.minor = element_blank(),
      axis.text.x = element_text(angle = 45, vjust = 1.0, hjust=1),
      legend.position="none") +
  scale_fill_manual(values=c("#ef476f", "#073b4c", "#ffd166")) +
  ggtitle("Figure 10. Box plots of log2 TER \nbetween original MDA vs GTC-stalled MDA samples.\nThe p-values were produced by t tests comparing log2 TERs \nfrom Xtail with those from naive Ribolog or Ribolog with CELP.")
```




